serial.ts

```diff
@@ -53,119 +81,291 @@ const kParityIndexMapping: ParityType[] =
+      }
       throw new Error('Error setting up device: ' + error.toString());
     }
   }
 
+  /**
+   * Closes the port.
+   *
+   * @return {Promise<void>} A promise that will resolve when the port is
+   * closed.
+   */
+  public async close(): Promise<void> {
+    await Promise.all([this.readable.cancel(), this.writable.abort()]);
+    await this.setSignals({dtr: false});
+    await this.device_.close();
+  }
+
   /**
    * A function used the get the options directoly from the device
-   * @return {Promise} A promise that will resolve with an object containing
-   * the device serial options
+   * @return {Promise<SerialOptions>} A promise that will resolve with an object
+   * containing the device serial options
    */
-  public async getInfo() {
+  public async getInfo(): Promise<SerialOptions> {
     const response = await this.device_.controlTransferIn({
       'requestType': 'class',
       'recipient': 'interface',
       'request': kGetLineCoding,
       'value': 0x00,
-      'index': 0x00,
+      'index': this.controlInterface_.interfaceNumber,
     }, 7);
 
-    if (response.status === 'ok') {
-      return this.readLineCoding(response.data!.buffer);
+    if (response.status === 'ok' && response.data) {
+      return this.readLineCoding(response.data.buffer);
     }
+    throw new DOMException('NetworkError', 'Failed to get serial line coding.');
   }
 
   /**
    * A function used to change the serial settings of the device
    * @param {object} options the object which carries serial settings data
-   * @return {Promise} A promise that will resolve when the options are set
+   * @return {Promise<void>} A promise that will resolve when the options are
+   * set
    */
-  public setOptions(options?: SerialOptions) {
-    const newOptions = {...this.serialOptions_, ...options};
-    this.serialOptions_ = newOptions;
+  public reconfigure(options: SerialOptions): Promise<void> {
+    this.serialOptions_ = {...this.serialOptions_, ...options};
     this.validateOptions();
-    return this.setSerialOptions();
+    return this.setLineCoding();
   }
 
   /**
-   * Set the device inside the class and figure out which interface is the
-   * proper one for transfer and control
-   * @param {object} device A device acquired from the WebUSB API
+   * Sets control signal state for the port.
+   * @param {SerialOutputSignals} signals The signals to enable or disable.
+   * @return {Promise<void>} a promise that is resolved when the signal state
+   * has been changed.
    */
-  private setPort(device: USBDevice) {
-    if (!SerialPort.isSerialDevice(device)) {
-      throw new TypeError('This is not a serial port');
+  public async setSignals(signals: SerialOutputSignals): Promise<void> {
+    this.outputSignals_ = {...this.outputSignals_, ...signals};
+
+    if (signals.dtr !== undefined || signals.rts !== undefined) {
+      const value = (this.outputSignals_.dtr ? 1 << 0 : 0) |
+                    (this.outputSignals_.rts ? 1 << 1 : 0);
```

Comment:
I think this calculation deserves a bit of explanation/background.  Maybe just a link to s spec or something?

Reply:
I've added links to the specification.